app/models/project.rb

```diff
@@ -38,6 +38,9 @@ class Project < ApplicationRecord
   has_one :readme, through: :repository
   has_many :repository_maintenance_stats, through: :repository
 
+  scope :least_recently_updated_stats, -> { joins(:repository_maintenance_stats).group('projects.id').where.not(repository: nil).order(Arel.sql('max(repository_maintenance_stats.updated_at) ASC')) }
```

Comment:
This query should find project's with stats, find the latest updated date for that project's stats, and then return them in ascending order. The idea is to find projects that have stats already and find the ones with the oldest stats so that they can be refreshed.

`Arel.sql()` was added to get rid of some deprecation warnings that are shown from rails.

Reply:
Explain:

```
 QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=1886.36..1889.40 rows=1214 width=417)
   Sort Key: (max(repository_maintenance_stats.updated_at))
   ->  HashAggregate  (cost=1812.03..1824.17 rows=1214 width=417)
         Group Key: projects.id
         ->  Merge Join  (cost=207.52..1805.96 rows=1214 width=417)
               Merge Cond: (repositories.id = projects.repository_id)
               ->  Merge Join  (cost=0.57..2782.16 rows=27024 width=20)
                     Merge Cond: (repositories.id = repository_maintenance_stats.repository_id)
                     ->  Index Only Scan using repositories_pkey on repositories  (cost=0.29..1483.00 rows=25298 width=4)
                     ->  Index Scan using index_repository_maintenance_stats_on_repository_id on repository_maintenance_stats  (cost=0.29..2195.42 rows=27024 width=16)
               ->  Sort  (cost=206.94..210.69 rows=1500 width=409)
                     Sort Key: projects.repository_id
                     ->  Seq Scan on projects  (cost=0.00..127.81 rows=1500 width=409)
                           Filter: (repository_id IS NOT NULL)

```

Reply:
We should probably have the `repository_maintenance_stats.updated_at` indexed. I can try that locally to see if it affects that HashAggregate cost.

Reply:
The updated_at worries me a little less than the sequential scan on projects to get ones without repository_id.

Reply:
I think I can take out the part where it is looking for a non null value for `repository_id`. I thought that would help limit the extent of the query but may not be worth it given this plan result.

Reply:
This is without the `nil` check on `repository_id`:
```
QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=1950.91..1954.91 rows=1602 width=417)
   Sort Key: (max(repository_maintenance_stats.updated_at))
   ->  HashAggregate  (cost=1849.62..1865.64 rows=1602 width=417)
         Group Key: projects.id
         ->  Merge Join  (cost=236.86..1841.61 rows=1602 width=417)
               Merge Cond: (repositories.id = projects.repository_id)
               ->  Merge Join  (cost=0.57..2782.16 rows=27024 width=20)
                     Merge Cond: (repositories.id = repository_maintenance_stats.repository_id)
                     ->  Index Only Scan using repositories_pkey on repositories  (cost=0.29..1483.00 rows=25298 width=4)
                     ->  Index Scan using index_repository_maintenance_stats_on_repository_id on repository_maintenance_stats  (cost=0.29..2195.42 rows=27024 width=16)
               ->  Sort  (cost=236.29..241.24 rows=1981 width=409)
                     Sort Key: projects.repository_id
                     ->  Seq Scan on projects  (cost=0.00..127.81 rows=1981 width=409)
```

It looks like there is still a sequential scan of the projects table.